### ISSUE TITLE
Annotation to provide validation features on top of types

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/ballerina-spec#239]
We could get quite a lot of mileage from a simple @validate annotation with fields describing the kinds of validation to perform, e.g.

```
@validate { minInclusive = 1 };
type PositiveInt int;
```

For numeric types we could follow XML Schema part 2 and allow:

- minInclusive
- maxInclusive
- minExclusive
- maxExclusive

Other obvious things:

- maxLength (for anything with a length, so strings and lists)
- minLength (similarly)
- pattern/regexp (on strings)

A more general and sophisticated way of doing this would be refinement types (#238), but that is orders of magnitude more work.

This doesn't deal with cases where you really want to transform the data as well notably:

- date/time-related types including durations
- strings that represents lists using whitespace or other delimiter to delimit items (as with XML Schema 2 list datatype)

**Comment by jclark** [comment]
This can also be useful in conjunction with automated data mapping.

**Comment by sanjiva** [comment]
+1.

**Comment by srinathperera** [comment]
Looking from just data mapping point of view, statistics such as mean, standard deviation etc also useful ( for numerical fields). 

Also, another angle is that although user does not provide this, in Choreo, we can trace, measure, and annotate in the background.

**Comment by jclark** [comment]
This is potentially relevant to GraphQL. Simple types with validation constraints can be exposed as user-defined scalars.

**Comment by jclark** [comment]
I think we should perhaps do this in a more first-class way using refinement types #238.

**Comment by jclark** [comment]
@shafreenAnfar mentions JSR 303 https://beanvalidation.org/1.0/spec/

**Comment by shafreenAnfar** [comment]
Yes. And Bean Validation 2.0 is available under https://beanvalidation.org/2.0/spec/ (formerly JSR 380). I feel this is an integral part of writing REST APIs with Spring Boot.

The list of built-in constraints can be found [here](https://beanvalidation.org/2.0/spec/#builtinconstraints).
